File: problems/leetcode/42.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() < 2) {
            return 0;  
        }
        int left = 0, right = height.size() - 1, volume = 0, max_left = height[left], max_right = height[right];
        while (left < right) {
            if (max_right < max_left) {
                right--;
                max_right = max(max_right, height[right]);
                volume += abs(max_right - height[right]);
            } else {
                left++;
                max_left = max(max_left, height[left]);
                volume += abs(max_left - height[left]);
            }
        }
        return volume;
    }
};

File: problems/leetcode/42.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) < 2:
            return 0
        left, right = 0, len(height) - 1
        volume = 0
        maxleft, maxright = height[left], height[right]

        while right > left:
            if maxright < maxleft:
                right -= 1
                maxright = max(height[right], maxright)
                volume += abs(maxright - height[right])
            else:
                left += 1
                maxleft = max(height[left], maxleft)
                volume += abs(maxleft - height[left])
        return volume
